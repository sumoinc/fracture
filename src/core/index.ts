export * from "./access-pattern";
export * from "./account";
export * from "./audit-strategy";
export * from "./fracture";
export * from "./index";
export * from "./naming-strategy";
export * from "./operation";
export * from "./organization";
export * from "./organizational-unit";
export * from "./region";
export * from "./resource-attribute";
export * from "./resource";
export * from "./service";
export * from "./structure";
