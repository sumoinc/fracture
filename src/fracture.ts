import { Project } from 'projen';
import { AppSync } from './appsync/appsync';
import { FractureComponent } from './component';
import { Entity, EntityOptions } from './model';

export interface FractureOptions {
  /**
   * Directory where generated code will be placed.
   */
  gendir?: string;
  /**
   * Enable AppSync
   */
  appsync?: boolean;
}

export class Fracture extends FractureComponent {
  public readonly gendir: string;
  public readonly appsync: boolean;

  constructor(
    project: Project,
    namespace?: string,
    options: FractureOptions = {},
  ) {
    super(project, namespace ?? 'fracture' );

    this.gendir = options.gendir ?? project.outdir;
    this.appsync = options.appsync ?? true;

    if (this.appsync) {
      new AppSync(project, this.namespace, { gendir: this.gendir });
    }
  }

  /**
   * Get all entities in this project's namespace.
   */
  public get entities(): Entity[] {
    const isEntity = (c: FractureComponent): c is Entity => c instanceof Entity && c.namespace === this.namespace;
    return (this.project.components as FractureComponent[]).filter(isEntity);
  }

  public addEntity(options: EntityOptions) {
    return new Entity(this.project, this.namespace, options);
  }


}
