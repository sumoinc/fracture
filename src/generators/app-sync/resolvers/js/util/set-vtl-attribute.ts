import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
