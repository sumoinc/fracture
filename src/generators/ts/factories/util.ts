import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
