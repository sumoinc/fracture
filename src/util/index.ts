export * from "./determineBuildId";
export * from "./test-util";
