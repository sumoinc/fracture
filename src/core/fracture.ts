import { Component, LoggerOptions, Project, LogLevel } from "projen";
import { deepMerge } from "projen/lib/util";
import { AuditStrategy } from "./audit-strategy";
import { NamingStrategy, NAMING_STRATEGY_TYPE } from "./naming-strategy";
import { OPERATION_SUB_TYPE } from "./operation";
import { Organization, OrganizationOptions } from "./organization";
import {
  ResourceAttributeGenerator,
  ResourceAttributeType,
} from "./resource-attribute";
import { Service, ServiceOptions } from "./service";
import { STRUCTURE_TYPE } from "./structure";
import { TypescriptService } from "../generators/ts/typescript-service";

/**
 *
 * Top level options for Fracture. These options also apply to services and
 * other sub components as they are created, unless they are overridden.
 *
 */
export interface FractureOptions {
  /**
   * Directory where generated code will be placed.
   * @default project.outdir + "/" + namespace
   */
  outdir?: string;
  /**
   * Logging options
   * @default LogLevel.INFO
   */
  logging?: LoggerOptions;
  /**
   * Versioned.
   * @default true
   */
  isVersioned?: boolean;
  /**
   * The naming strategy to use for generated code.
   */
  namingStrategy?: NamingStrategy;
  /**
   * The audit strategy to use for generated code.
   */
  auditStrategy?: AuditStrategy;
}

/**
 * The root of the entire application.
 */
export class Fracture extends Component {
  // member components
  public readonly services: Service[];
  public readonly organizations: Organization[];
  // project and namespace
  public readonly project: Project;
  public readonly namespace: string;
  // all other options
  public readonly options: Required<FractureOptions>;

  constructor(
    project: Project,
    namespace: string = "fracture",
    options: FractureOptions = {}
  ) {
    super(project);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     * These are the options that will be used through all code generation
     * unless explicitly overridden.
     *
     **************************************************************************/

    const defaultOptions: Required<FractureOptions> = {
      outdir: namespace,
      logging: {
        level: LogLevel.INFO,
      },
      isVersioned: true,
      namingStrategy: {
        ts: {
          attributeName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          className: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          enumName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          fileName: NAMING_STRATEGY_TYPE.PARAM_CASE,
          functionName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          functionParameterName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          interfaceName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          typeName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
        },
        appsync: {
          vtl: {
            file: NAMING_STRATEGY_TYPE.PARAM_CASE,
          },
        },
        attributes: {
          compositionSeperator: "#",
        },
        structures: {
          prefixes: {
            [STRUCTURE_TYPE.DATA]: "",
            [STRUCTURE_TYPE.INPUT]: "",
            [STRUCTURE_TYPE.OUTPUT]: "",
            [STRUCTURE_TYPE.TRANSIENT]: "",
          },
          suffixes: {
            [STRUCTURE_TYPE.DATA]: "",
            [STRUCTURE_TYPE.INPUT]: "input",
            [STRUCTURE_TYPE.OUTPUT]: "output",
            [STRUCTURE_TYPE.TRANSIENT]: "message",
          },
        },
        operations: {
          prefixes: {
            [OPERATION_SUB_TYPE.CREATE_ONE]: "create",
            [OPERATION_SUB_TYPE.READ_ONE]: "get",
            [OPERATION_SUB_TYPE.UPDATE_ONE]: "update",
            [OPERATION_SUB_TYPE.DELETE_ONE]: "delete",
            [OPERATION_SUB_TYPE.IMPORT_ONE]: "import",
            [OPERATION_SUB_TYPE.LIST]: "list",
          },
          suffixes: {
            [OPERATION_SUB_TYPE.CREATE_ONE]: "",
            [OPERATION_SUB_TYPE.READ_ONE]: "",
            [OPERATION_SUB_TYPE.UPDATE_ONE]: "",
            [OPERATION_SUB_TYPE.DELETE_ONE]: "",
            [OPERATION_SUB_TYPE.IMPORT_ONE]: "",
            [OPERATION_SUB_TYPE.LIST]: "",
          },
        },
      },
      auditStrategy: {
        create: {
          dateAttribute: {
            name: "created-at",
            shortName: "cd",
            comments: [`The date and time this record was created.`],
            type: ResourceAttributeType.DATE_TIME,
            isRequired: true,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [OPERATION_SUB_TYPE.CREATE_ONE],
          },
        },
        update: {
          dateAttribute: {
            name: "updated-at",
            shortName: "ud",
            comments: [`The date and time this record was last updated.`],
            type: ResourceAttributeType.DATE_TIME,
            isRequired: true,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [
              OPERATION_SUB_TYPE.CREATE_ONE,
              OPERATION_SUB_TYPE.UPDATE_ONE,
              OPERATION_SUB_TYPE.DELETE_ONE,
            ],
          },
        },
        delete: {
          dateAttribute: {
            name: "deleted-at",
            shortName: "dd",
            comments: [`The date and time this record was deleted.`],
            type: ResourceAttributeType.DATE_TIME,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [OPERATION_SUB_TYPE.DELETE_ONE],
            outputOn: [OPERATION_SUB_TYPE.DELETE_ONE],
          },
        },
      },
    };

    /***************************************************************************
     *
     * INIT FRACTURE PROJECT
     *
     **************************************************************************/

    // member components
    this.services = [];
    this.organizations = [];

    // project and namespace
    this.project = project;
    this.namespace = namespace;

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<FractureOptions>;

    return this;
  }

  /**
   * Build the project.
   *
   * Call this when you've configured everything, prior to preSynthesize.
   *
   * @returns void
   */
  public build() {
    this.services.forEach((service) => {
      service.build();
      new TypescriptService(service);
    });
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add a service to the fracture project.
   * @param {ServiceOptions}
   * @returns {Service}
   */
  public addService(options: ServiceOptions) {
    return new Service(this, options);
  }

  /**
   * Add an organization to the fracture project.
   * @param {OrganizationOptions}
   * @returns {Organization}
   */
  public addOrganization(options: OrganizationOptions) {
    return new Organization(this, options);
  }
}
