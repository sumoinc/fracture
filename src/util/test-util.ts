import { Project } from "projen";
import { synthSnapshot } from "projen/lib/util/synth";

export const TEST_ACCOUNT_ONE = "000000000000";
export const TEST_ORG_ONE = "org-123456";
export const TEST_REGION_ONE = "us-east-1";

/**
 *
 * Helper to generate and return one file from fracture.
 * Useful when trying to test the final output of generated files.
 *
 * @param fracture
 * @param filepath
 * @returns collection of files starting with supplied filepath
 */
export const synthFiles = (project: Project, filepath: string = ""): any => {
  const snapshot = synthSnapshot(project);
  // console.log(Object.keys(snapshot));
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(filepath))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {} as { [key: string]: any });
  return filtered;
};

export const synthFile = (project: Project, filepath: string): string => {
  const files = synthFiles(project, filepath);

  if (files.length > 1) {
    throw new Error(`more than one file matched filepath "${filepath}"`);
  }

  if (files.length === 0) {
    throw new Error(`No files matched filepath "${filepath}"`);
  }

  return files[filepath];
};

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testProject = () => {
//   return new Project({
//     name: "my-project",
//   });
// };

// export const testNodeProject = (options: Partial<NodeProjectOptions> = {}) => {
//   return new NodeProject({
//     name: "my-project",
//     defaultReleaseBranch: "main",
//     packageManager: NodePackageManager.PNPM,
//     pnpmVersion: "8",
//     ...options,
//   });
// };

// /**
//  * Builds a simple typescript app project as a test harness
//  */
// export const testFractureProject = (
//   options: Partial<FractureProjectOptions> = {}
// ) => {
//   const fp = new FractureProject({
//     name: "my-project",
//     ...options,
//   });

//   /*
//   fp.buildEnvironment = new AwsEnvironment(fp, {
//     name: "test",
//     account: TEST_ACCOUNT_ONE,
//     region: TEST_REGION_ONE,
//     primaryHostedZoneId: "foo",
//     primaryDomainName: "bar",
//   });
//   */

//   return fp;
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testApp = () => {
//   return new App({
//     parent: testFractureProject(),
//     name: "my-app",
//   });
// };

// /**
//  * Builds a simple data service as a test harness
//  */
// export const testDataService = () => {
//   return new DataService(testApp(), {
//     name: "my-service",
//   });
// };

// /**
//  * Build a simple resource to test with
//  */
// export const testResource = () => {
//   const service = testDataService();
//   return new Resource(service.app, {
//     service,
//     name: "my-resource",
//   });
// };

// /**
//  * A more complicated resource configuration
//  */
// export const complexService = () => {
//   const service = testDataService();

//   const person = new Resource(service.app, { service, name: "person" });

//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     comments: ["A club"],
//     attributeOptions: [{ name: "name" }],
//   });

//   person.addAttribute({
//     name: "first-name",
//     shortName: "fn",
//   });
//   person.addAttribute({
//     name: "last-name",
//     shortName: "ln",
//   });
//   person.addAttribute({
//     name: "height",
//     shortName: "ht",
//     type: ResourceAttributeType.INT,
//   });
//   person.addAttribute({
//     name: "is-frog",
//     shortName: "if",
//     type: ResourceAttributeType.BOOLEAN,
//     comments: ["Is this person a frog?"],
//   });
//   person.addAttribute({
//     name: "fav-colors",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: ResourceAttributeType.STRING,
//     comments: ["List opf favorite colors"],
//   });

//   // one specific club
//   person.addAttribute({
//     name: "one-club",
//     type: club,
//   });
//   // an array of clubs
//   person.addAttribute({
//     name: "array-of-clubs",
//     shortName: "ad",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });

//   // an array of persons
//   club.addAttribute({
//     name: "array-of-persons",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });

//   // operation
//   Operation.create(person);
//   Operation.read(person);
//   Operation.update(person);
//   Operation.delete(person);

//   return service;
// };

// /**
//  * Build a simple resource attribute to test with
//  */
// export const testResourceAttribute = () => {
//   const service = testDataService();
//   const resource = testResource();
//   return new ResourceAttribute(service.app, {
//     name: "my-attr",
//     resource,
//   });
// };

// /**
//  * Build a simple structure to test with
//  */
// export const testStructure = () => {
//   const service = testDataService();
//   return new Structure(service.app, { name: "my-structure" });
// };

// /**
//  * Build a simple operation to test with
//  */
// export const testOperation = () => {
//   const resource = testResource();
//   return new Operation(resource.service.app, {
//     name: "my-operation",
//     resource,
//   });
// };

// /**
//  * Build a nuxt site to test with
//  */
// export const testNuxtJsSite = () => {
//   return new NuxtJsSite({
//     parent: testFractureProject(),
//     name: "my-site",
//   });
// };
// export const testServiceTwoRelatedResources = () => {
//   const service = testDataService();
//   const person = new Resource(service.app, {
//     service,
//     name: "person",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   const club = new Resource(service.app, {
//     service,
//     name: "club",
//     attributeOptions: [{ name: "name", shortName: "n" }],
//   });
//   club.addAttribute({
//     name: "people",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: person,
//   });
//   person.addAttribute({
//     name: "clubs",
//     type: ResourceAttributeType.ARRAY,
//     typeParameter: club,
//   });
//   return { service, person, club };
// };
