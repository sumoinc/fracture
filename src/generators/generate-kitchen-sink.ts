import { Component } from "projen";
import { GeneratedTypes } from "./ts";
import { GeneratedCdkApp } from "./ts/cdk-app";
import { GeneratedConfigurations } from "./ts/configurations";
import { Fracture, FractureService } from "../core";

export class GenerateKitchenSink extends Component {
  constructor(fracture: Fracture) {
    super(fracture);

    /***************************************************************************
     * Typescript Service Generators
     **************************************************************************/

    FractureService.all(fracture).forEach((service) => {
      new GeneratedTypes(service);
      new GeneratedCdkApp(service);
      new GeneratedConfigurations(service);
    });
  }
}
