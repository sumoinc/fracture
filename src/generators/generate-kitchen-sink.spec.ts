import { GenerateKitchenSink } from "./generate-kitchen-sink";
import { FractureService } from "../core";
import { Fracture } from "../core/fracture.ts-disabled";
import { synthFiles } from "../util/test-util";

describe("success conditions", () => {
  test("Smoke test", () => {
    const fracture = new Fracture();
    const service = new FractureService(fracture, {
      name: "my-service",
    });
    expect(service).toBeTruthy();

    new GenerateKitchenSink(fracture);

    // snapshot all files.
    const content = synthFiles(fracture, "");
    const fileList = Object.keys(content);
    expect(fileList).toMatchSnapshot();

    //console.log(fileList);
  });
});
