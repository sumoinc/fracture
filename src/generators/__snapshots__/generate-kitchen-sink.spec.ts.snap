// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success conditions Smoke test 1`] = `
Array [
  ".eslintrc.json",
  ".gitattributes",
  ".github/pull_request_template.md",
  ".github/workflows/build.yml",
  ".github/workflows/deploy-chore.yml",
  ".github/workflows/deploy-feature.yml",
  ".github/workflows/deploy-fix.yml",
  ".github/workflows/deploy-main.yml",
  ".github/workflows/pull-request-lint.yml",
  ".github/workflows/release.yml",
  ".github/workflows/upgrade-main.yml",
  ".gitignore",
  ".mergify.yml",
  ".npmignore",
  ".npmrc",
  ".prettierignore",
  ".prettierrc.json",
  ".projen/deps.json",
  ".projen/files.json",
  ".projen/tasks.json",
  ".vscode/extensions.json",
  ".vscode/settings.json",
  "package.json",
  "pnpm-workspace.yaml",
  "README.md",
  "services/my-service/.eslintrc.json",
  "services/my-service/.gitattributes",
  "services/my-service/.gitignore",
  "services/my-service/.npmignore",
  "services/my-service/.npmrc",
  "services/my-service/.prettierignore",
  "services/my-service/.prettierrc.json",
  "services/my-service/.projen/deps.json",
  "services/my-service/.projen/files.json",
  "services/my-service/.projen/tasks.json",
  "services/my-service/cdk.json",
  "services/my-service/package.json",
  "services/my-service/README.md",
  "services/my-service/src/generated/configurations.ts",
  "services/my-service/src/generated/main.ts",
  "services/my-service/src/generated/types.ts",
  "services/my-service/src/main.ts",
  "services/my-service/test/main.test.ts",
  "services/my-service/tsconfig.dev.json",
  "services/my-service/tsconfig.json",
  "src/index.ts",
  "test/hello.test.ts",
  "tsconfig.dev.json",
  "tsconfig.json",
  "turbo.json",
]
`;
